packages/atclient/src/atclient.c

```diff
@@ -239,3 +243,300 @@ void atclient_free(atclient *ctx) {
   atclient_connection_free(&(ctx->secondary_connection));
 }
+
+static int atclient_create_shared_encryption_key(atclient *ctx, const atclient_atsign *recipient,
+                                                 char *enc_key_shared_by_me);
+
+static int atclient_get_public_encryption_key(atclient *ctx, const atclient_atsign *atsign,
+                                              char *public_encryption_key);
+
+int atclient_get_encryption_key_shared_by_me(atclient *ctx, const atclient_atsign *recipient,
+                                             char *enc_key_shared_by_me) {
+  int ret = 1;
+
+  // llookup:shared_key.recipient_atsign@myatsign
+  char *command_prefix = "llookup:shared_key.";
+  short command_prefix_len = 19;
+  short atsign_with_at_len = (short)strlen(ctx->atsign.atsign);
+
+  short command_len = command_prefix_len + (atsign_with_at_len * 2 - 1) + 3;
+  char command[command_len];
+  snprintf(command, command_len, "llookup:shared_key.%s%s\r\n", recipient->without_prefix_str, ctx->atsign.atsign);
+
+  const unsigned long recvlen = 1024;
+  unsigned char recv[sizeof(unsigned char) * recvlen];
+  memset(recv, 0, sizeof(unsigned char) * recvlen);
+  unsigned long olen = 0;
+
+  ret = atclient_connection_send(&(ctx->secondary_connection), command, strlen((char *)command), recv, recvlen, &olen);
+  if (ret != 0) {
+    return ret;
+  }
+
+  char *response = recv;
+
+  // Truncate response: "@" + myatsign + "@"
+  int response_prefix_len = atsign_with_at_len + 2;
+  char response_prefix[response_prefix_len];
+  snprintf(response_prefix, response_prefix_len, "@%s@", ctx->atsign.without_prefix_str);
+
+  if (atclient_stringutils_starts_with(response, recvlen, response_prefix, response_prefix_len)) {
+    response = response + response_prefix_len;
+  }
+
+  if (atclient_stringutils_ends_with(response, recvlen, response_prefix, response_prefix_len)) {
+    response[strlen(response) - response_prefix_len - 1] = '\0';
+  }
+
+  // does my atSign already have the recipient's shared key?
+  if (atclient_stringutils_starts_with(response, recvlen, "data:", 5)) {
+    response = response + 5;
+
+    // 44 + 1
+    unsigned long plaintextlen = 45;
+    unsigned char plaintext[sizeof(unsigned char) * plaintextlen];
+    memset(plaintext, 0, plaintextlen);
+    unsigned long plaintextolen = 0;
+
+    ret = atchops_rsa_decrypt(ctx->atkeys.encryptprivatekey, (const unsigned char *)response, strlen((char *)response),
+                              plaintext, plaintextlen, &plaintextolen);
+    if (ret != 0) {
+      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atchops_rsa_decrypt: %d\n", ret);
+      return ret;
+    }
+
+    memcpy(enc_key_shared_by_me, plaintext, plaintextlen);
+  }
+
+  else if (atclient_stringutils_starts_with(recv, recvlen, "error:AT0015-key not found",
+                                            strlen("error:AT0015-key not found"))) {
+    // TODO: or do I need to create, store and share a new shared key?
```

Comment:
I don't think we need to create / share a new shared key, what if we just want to check whether a key has been shared or not.

We should have a special return code to allow us to identify this situtation.

Reply:
This is what all our other SDKs do. I know C focus is different, but I think it's the way the method should be working. Is there any other reason (besides obtaining it or creating it if it does not exist) why a developer would want to know whether or not a key has already been created? We can introduce an extra opcode/bool parameter to let the user decide if there's a common situation on which it is helpful

Reply:
Yeah, we don't need things to be so automatic in the C sdk. We should focus on things being simple and relinquishing control to the application developer. It will help us keep things light weight and easier to maintain.